### PR TITLE
WIP: DB connections migration

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates all DB connections to the existing DBAL library. `DBCore` is kept for backward compatibity and is used as a proxy.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | Yes, all subclasses of DBCore
| Deprecations? | Yes, all subclasses of DBCore
| Fixed ticket? | /
| How to test?  | All tests requiring db connexions must works, as well as your front office and back-office

### CODE REVIEW

File: classes/db/Db.php [hunk 1]
```diff
@@ -94,91 +99,195 @@
...
      *
      * @param string $str
      * @return string
      */
-    abstract public function _escape($str);
+    public function _escape($str)
+    {
+        $search = array("\\", "\0", "\n", "\r", "\x1a", "'", '"');
+        $replace = array("\\\\", "\\0", "\\n", "\\r", "\Z", "\'", '\"');
+        return str_replace($search, $replace, $str);    }
```
Comment (jocel1) [inline classes/db/Db.php]: better use $this->link->quote();

File: classes/db/Db.php [hunk 2]
```diff
@@ -187,14 +296,11 @@
...
-    /**
-     * Selects best table engine.
-     *
-     * @return string
-     */
-    abstract public function getBestEngine();
+    public function set_db($db_name)
+    {
+        // ToDo: To be checked
+        throw new Symfony\Component\Serializer\Exception\UnsupportedException();
```
Comment (jocel1) [inline classes/db/Db.php]: $this->query("use ".$dbName); ?

File: classes/db/Db.php [hunk 3]
```diff
@@ -187,14 +296,11 @@
...
      */
-    abstract public function set_db($db_name);
-
-    /**
-     * Selects best table engine.
-     *
-     * @return string
-     */
-    abstract public function getBestEngine();
+    public function set_db($db_name)
```
Comment (jocel1) [inline classes/db/Db.php]: camelCase ?

File: classes/db/Db.php [hunk 4]
```diff
@@ -777,6 +798,49 @@ public function escape($string, $html_ok = false, $bq_sql = false)
     }
 
     /**
+     * Get used link instance
+     *
+     * @return PDO|mysqli|resource Resource
```
Comment (jocel1) [inline classes/db/Db.php]: @return \Doctrine\DBAL\Connection ?

File: classes/db/Db.php [hunk 5]
```diff
@@ -790,53 +854,83 @@ public function escape($string, $html_ok = false, $bq_sql = false)
...
      * @param string $pwd Password for database connection
-     * @return bool
+     * @param string $db Database name
+     * @param bool $newDbLink
+     * @param string|bool $engine
+     * @param int $timeout
+     * @return int Error code or 0 if connection was successful
      */
-    public static function checkEncoding($server, $user, $pwd)
+    public static function tryToConnect($server, $user, $pwd, $db, $new_db_link = true, $engine = null, $timeout = 5)
```
Comment (jocel1) [inline classes/db/Db.php]: camelCase for new variables

File: classes/db/Db.php [hunk 6]
```diff
@@ -790,53 +854,83 @@ public function escape($string, $html_ok = false, $bq_sql = false)
...
+        if (!$result) {
+            $value = 'MyISAM';
+        }else {
+            $row = $result->fetch();
+            if (!$row || strtolower($row['Value']) != 'yes') {
+                $value = 'MyISAM';
+            }
+        }
+        /* MySQL >= 5.6 */
+        $sql = 'SHOW ENGINES';
```
Comment (jocel1) [inline classes/db/Db.php:878]: FYI SELECT ENGINE FROM INFORMATION_SCHEMA.ENGINES; covers MySQL >= 5.0.
Not needed anymore if SHOW VARIABLES LIKE 'innodb_%'; is used

File: classes/db/Db.php [hunk 7]
```diff
@@ -790,53 +854,83 @@ public function escape($string, $html_ok = false, $bq_sql = false)
...
-     * @param string $prefix Tables prefix
-     * @return bool
+     * @return string
      */
-    public static function hasTableWithSamePrefix($server, $user, $pwd, $db, $prefix)
+    public function getBestEngine()
     {
-        return call_user_func_array(array(Db::getClass(), 'hasTableWithSamePrefix'), array($server, $user, $pwd, $db, $prefix));
+        $value = 'InnoDB';
+        $sql = 'SHOW VARIABLES WHERE Variable_name = \'have_innodb\'';
```
Comment (jocel1) [inline classes/db/Db.php]: SHOW VARIABLES LIKE 'innodb_%'; to cover old version ?

**Review by jocel1 — CHANGES_REQUESTED** [review]

File: classes/db/Db.php (hunk 3 above)
Comment (Quetzacoalt91) [reply classes/db/Db.php]: For backward compatibility reasons, I won't change the function name.

File: classes/db/Db.php (hunk 2 above)
Comment (Quetzacoalt91) [reply classes/db/Db.php]: Yeah obviously. Will be updated.

File: classes/db/Db.php (hunk 1 above)
Comment (Quetzacoalt91) [reply classes/db/Db.php]: It seems quote() is used for a specific variable before being inserted in a sql request. With _escape, we search in the  whole sql request .

**Comment by Quetzacoalt91** [comment]
To be done later, when we will be able to have the symfony kernel during Db class instanciation.

One of the objective of this PR is to share the DB connection between Symfony and the legacy. This requires the Symfony kernel to be initialized BEFORE the legacy one. Actually, this is not possible even in the back-office, so I lose this performance gain.

I close this PR until we get a Symfony kernel available earlier, and this must be done somewhere else.

**Pull request closed without merging**